File: Algorithms Lib/practice_problems/failures/save_nagato.cpp
// Still TLE

#include <iostream>
#include <vector>
#include <tuple>
using namespace std;
#define MAX 500001
#define vi vector<int>
#define pb push_back
#include <algorithm>
vi graph[MAX];
vi children[MAX];
vi parents[MAX];
int vis[MAX] = {0};
int downmemo[MAX] = {0};
int upmemo[MAX] = {0};
int downmemo2[MAX] = {0};
int bestused[MAX] = {0};

void dfs(int n){
    vis[n] = 1;
    for (int node : graph[n]){
        if (vis[node] == 0){
            children[n].pb(node);
            parents[node].pb(n);
            dfs(node);
        }
    }
}

tuple<int, int> downdp(int k){
    int best = 0;
    int secondbest = 0;
    if (downmemo[k] != 0 && downmemo2[k] != 0){
        return make_tuple(downmemo[k], downmemo2[k]);
    }
    for (int node : children[k]){
        auto [ans, useless] = downdp(node);
        ans += 1;
        if (ans > best){
            if (best > secondbest){
                secondbest = best;
            }
            best = ans;
            bestused[k] = node;
        }
        else {
            if (ans > secondbest){
                secondbest = best;
            }
        }
    }
    downmemo[k] = best;
    downmemo2[k] = secondbest;
    return make_tuple(best, secondbest);
}
int updp(int k){
    int up = 0;
    if (upmemo[k] != 0){
        return upmemo[k];
    }
    for (auto parent : parents[k]){
        auto [down1, down2] = downdp(parent);
        if (bestused[parent] == k){
            up = down2+1;
        }
        else {
            up = down1+1;
        }
        up = max(up, 1 + updp(parent));
    }
    upmemo[k] = up;
    return up;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    cin >> n;
    for (int i = 0; i < n-1; i++){
        int u; int v;
        cin >> u;
        cin >> v;
        graph[u].pb(v);
        graph[v].pb(u);
    }
    dfs(1);
    for (int i=1; i<=n; i++){
        auto [x, y] = downdp(i);
        cout << max(x, updp(i))+1 << "\n";
    }
}

File: Algorithms Lib/practice_problems/failures/save_nagato.py
# TLEs... and MLEs... I think the Time Complexity is correct though

import sys
input = sys.stdin.readline
print = sys.stdout.write
sys.setrecursionlimit(10000)

n = int(input())
graph = [[] for i in range(n+1)]
children = [[] for i in range(n+1)]
parents = [[] for i in range(n+1)]
visited = [0 for i in range(n+1)]
memo = [[None,None,None] for i in range(n+1)]
bestused = [0 for i in range(n+1)]

for i in range(n-1):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

def dfs(n):
    visited[n] = 1 # Mark node as visited
    for node in graph[n]: 
        if(visited[node] == 0): # Check next unvisited node
            children[n].append(node)
            parents[node].append(n)
            dfs(node)
dfs(1)

def downdp(k):
    best = 0
    secondbest = 0
    if memo[k][0] != None and memo[k][2] != None:
        return [memo[k][0], memo[k][2]]

    for node in children[k]:
        ans = downdp(node)[0]+1
        if ans > best:
            if best > secondbest:
                secondbest = best
            best = ans
            bestused[k] = node
        else:
            if ans > secondbest:
                secondbest = ans
    memo[k][0] = best
    memo[k][2] = secondbest
    return [best, secondbest]

def updp(k):
    up = 0
    if memo[k][1] != None:
        return memo[k][1]
    for parent in parents[k]:
        if bestused[parent] == k:
            up = downdp(parent)[1] + 1
        else:
            up = downdp(parent)[0] + 1
        up = max(up, 1 + updp(parent))
    memo[k][1] = up
    return up

for i in range(1, n+1):
    print(str(max(downdp(i)[0], updp(i))+1))
    print("\n")